GabrielPedroza/bread
# Webhook Implementation for GitHub

## Description

PR Accomplishments: This PR contains a working WebHook implementation for GitHub that is testable. 

Future PR Accomplishments: Future PR's will have the WebHook implementation for GitHub working for all users with side effects in a hosted project URL. This means it will also interact with a database to show that it has been created in the users home page.

## Milestones

This is a nice milestone as it is very close to having a working e2e implementation of the project. WebHooks are imperative for this project and this is a great start. 

This is a great milestone as WebHooks is a new feature that I've never implemented before and was able to not only learn WebHooks itself but have it working through an API. Even though a lot of code was not written on the project itself, a great chunk of time was spent learning and integrating WebHooks through varies methods (Cloud Computing: AWS Lambda & API Gateway, NGROK tunneling, Vercel WebHooks, Heroku, etc...).

## Test Plan

Under the `server/api/routers/webhook.ts` file, you would see these variables:

```ts
 const owner = "GabrielPedroza";
 const repo = "e-commerce";
```

1. Replace those strings with your GitHub username and an exisiting repo that you have. 

2. Press the create button to open the modal, click GitHub as event, and create.

3. Go to your repo and see the webhook under settings/webhooks.

4. You can also open the server (where you ran npm run dev) to see the response.

Screenshot proof of working webhook

1. When user creates an automation - https://ibb.co/hYkQjgc
2. Response of GitHub creating hook - https://ibb.co/1Z73sHW
3. GitHub Repo with created hook - https://ibb.co/2t6DGGr
5. If ran again, hooks already exists - https://ibb.co/xL0vN5s
6. If fake name or repo input (404) - https://ibb.co/d7DtZc9
